### PR TITLE
Anti spam-macro measure

### LINKED ISSUE #7994: you can still crash the server by radio spamming
**Problem Description**:
My wgw bind crashed the server

**What did you expect to happen**: 
Crash the server

**What happened instead**:
Crash the server

**Why is this bad/What are the consequences:**
CRASH THE SERVER

**Steps to reproduce the problem**:
Spam wgw with an alt + \ bind

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: CRASH THE SERVER

-- from the guy that crashed your server five minutes ago

### PR DESCRIPTION
**Pull request by Birdtalon**
Adds a 1 second cooldown to the say & whisper verb for all mobs. Shouldn't really affect normal play as most people won't be sending messages faster than once per second.

Hopefully will prevent any repeats of #7994 

Fixes #7994

### CODE REVIEW

File: code/modules/mob/living/say.dm [hunk 1]
```diff
@@ -359,6 +359,9 @@ proc/get_radio_key_from_channel(var/channel)
 	return src
 
 /mob/living/proc/whisper_say(var/message, var/datum/language/speaking = null, var/alt_name="", var/verb="whispers")
+	if(world.time < say_cooldown)
+		to_chat(src, "<span class='notice'>Take a breath before speaking again!</span>")
```
Comment (tigercat2000) [inline code/modules/mob/living/say.dm]: This proc can be called autonomously, please don't put any warnings in it.

File: code/modules/mob/say.dm [hunk 2]
```diff
@@ -25,7 +27,12 @@
 	*/
 
 	set_typing_indicator(0)
-	usr.say(message)
+	if(world.time > say_cooldown)
+		usr.say(message)
+		say_cooldown = world.time + 10 // We said something so take some breathing room.
+	else
+		to_chat(src, "<span class='notice'>Take a breath before speaking again!</span>") // Notify the user they've said things too fast.
```
Comment (tigercat2000) [inline code/modules/mob/say.dm]: Spammers spamming themselves can still lag the server by processing requests. Ideally you should just have `if(world.time > say_cooldown) return` above `usr.say(message)`.

**Comment by alexkar598** [comment]
Can you change it to half a second?

**Comment by SkeletalElite** [comment]
Yeah I feel like when an antag is getting me this anti spam can fuck me over yelling for help and yelling a location
Some locations take less than a second to type

**Comment by tigercat2000** [comment]
This should probably be closer to 1-5 deciseconds, just to stop crashing the server

File: code/modules/mob/living/say.dm (hunk 1 above)
Comment (Birdtalon) [reply code/modules/mob/living/say.dm]: Addressed

File: code/modules/mob/say.dm (hunk 2 above)
Comment (Birdtalon) [reply code/modules/mob/say.dm]: Addressed

**Comment by Birdtalon** [comment]
Review comments addressed.

- Removed user feedback.
- Cooldown reduced to 3 deciseconds.

**Comment by SamHPurp** [comment]
@tigercat2000 a lack of user feedback is bad design. People arre going to wonder where their messages are going and why they aren't being said, and it will do nothing but aggravate people.

**Comment by Ralta** [comment]
What @SamHPurp said. This was gonna be my question when I saw this PR - I already occasionally fuck up and type > instead of ? and the message doesn't go through - at least give some kind of indication that your message has failed

**Comment by Birdtalon** [comment]
I agree, however with the cooldown being now 30ms the incidences of a normal user typing a message in that time would be very minimal I expect.

**Comment by Ralta** [comment]
Fair point - that's a lot less likely to happen now, though still possible if you're like 
> MY VOXYGEN! IS DYING!
> SKREK!
> *cough
🐦

**Comment by taukausanake** [comment]
If you're able to type three anythings in one second then you're not human. You really shouldn't need some indicator at that point, but I guess one wouldn't hurt for robots who play SS13, so long as it doesn't cause another form of lag

**Comment by tigercat2000** [comment]
@taukausanake You'd actually be wrong, a very small series of messages, especially if you're intentionally splitting up a big message and/or doing the
each
line
for
effect
thing
can easily happen in less than one second.

**Comment by KingPhilipIII** [comment]
First my ability to flip endlessly and now this.

**Comment by Rumiluntti** [comment]
If
You
Type
Like
This
You
Deserve
Everything
Bad
Ever

**Comment by KasparoVy** [comment]
Started
from
the
bottom
now
we're
here

**Comment by Spacemanspark** [comment]
Could just have it mute someone automatically if they spam macros :^)

**Comment by Crazylemon64** [comment]
This *could* stand to be the type of cooldown, "more than X messages in the last Y seconds" - also, the indicator message of what went wrong would be a good idea.

**Comment by Fox-McCloud** [comment]
As I told you in coderchat, you need to pin down if this is causing server performance issues or client performance issues.

You also need to pin down if it's caused by just generally talking or if it's caused by talking over radio.

We want to cause as little impact as possible and also know exactly where the problem is coming from.

Also, as @Crazylemon64 said, that kind of cooldown method is preferable to this one.

**Comment by Birdtalon** [comment]
I don't really have time for that level of investigation right now I'm afraid, someone else is welcome to pick that up.

**Pull request closed without merging**